From b0cac696e36cbb1c0f59f8500e46780e0a6fc622 Mon Sep 17 00:00:00 2001
From: Daniel Emminizer <daniel.m.emminizer.civ@us.navy.mil>
Date: Mon, 17 Mar 2025 13:37:53 +0000
Subject: [PATCH 10/29] SDK: Append Path for protobuf Library (1/2)

**Release Notes:** CMake minimum version corrected to 3.21 from 3.20, for MSVC 2022 support.

**JIRA Issue:** SIM-18056, SIM-18058

**Description:** Now using APPEND_PATH for add_proto_library. This is required in order to generate the proto files in the right location in CMake older than about 3.27 and earlier.

**Notes:** This might be required in versions later than 3.27, but is not required in CMake 3.31 and later. It is not harmful to include in later versions.

**Testing Performed:** Configure and generate and build with MSVC 2022 on CMake 3.21.
---
 CMakeLists.txt                     |  4 ++--
 CMakeModules/AddProtoLibrary.cmake | 13 ++++++++++++-
 SDK/simData/CMakeLists.txt         |  1 +
 3 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index d4459e41..1a4b411b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,5 @@
-# MSVC 2022 support in 3.20
-cmake_minimum_required(VERSION 3.20)
+# MSVC 2022 support in 3.21
+cmake_minimum_required(VERSION 3.21)
 
 # Disable in-source builds and avoid writing to source directories directly
 set(CMAKE_DISABLE_SOURCE_CHANGES ON)
diff --git a/CMakeModules/AddProtoLibrary.cmake b/CMakeModules/AddProtoLibrary.cmake
index a6660002..d2fbb066 100644
--- a/CMakeModules/AddProtoLibrary.cmake
+++ b/CMakeModules/AddProtoLibrary.cmake
@@ -16,6 +16,7 @@ This is intended to replace the older CreateProtobufLibrary code.s
       <target-name>
       [STATIC|SHARED]
       [LITE]
+      [APPEND_PATH]
       [PROTO_DIR <dir>]
       [EXPORT_MACRO <macro>]
       [PROTOC_OUT_DIR <dir>]
@@ -40,6 +41,11 @@ This is intended to replace the older CreateProtobufLibrary code.s
   ``LITE``
     Link against the libprotobuf-lite library. Only valid for ``STATIC``
 
+  ``APPEND_PATH``
+    Pass-throguh to ``protobuf-generate``: A flag that causes the base path of all proto
+    schema files to be added to ``IMPORT_DIRS``. This may be useful when the proto files
+    are in subdirectories relative to the loaded CMakeLists.txt file.
+
   ``NO_HEADER``
     Do not generate the warning-free headers. By default, all headers are generated.
 
@@ -77,7 +83,7 @@ function(add_proto_library TARGETNAME)
         find_package(Protobuf REQUIRED)
     endif()
 
-    set(_options SHARED STATIC LITE NO_HEADER)
+    set(_options SHARED STATIC LITE NO_HEADER APPEND_PATH)
     set(_singleargs PROTO_DIR EXPORT_MACRO PROTO_OUT_DIR HDR_CODE_BLOCK)
     set(_multiargs PROTOC_OPTIONS IMPORT_DIRS)
     cmake_parse_arguments(arg "${_options}" "${_singleargs}" "${_multiargs}" ${ARGN})
@@ -93,6 +99,11 @@ function(add_proto_library TARGETNAME)
         LANGUAGE cpp
     )
 
+    # APPEND_PATH is a pass-through
+    if(arg_APPEND_PATH)
+        list(APPEND _GEN_ARGS APPEND_PATH)
+    endif()
+
     # Determine if static or shared; default to static
     set(_LIB_TYPE STATIC)
     if(arg_SHARED)
diff --git a/SDK/simData/CMakeLists.txt b/SDK/simData/CMakeLists.txt
index 4856450a..0002c748 100644
--- a/SDK/simData/CMakeLists.txt
+++ b/SDK/simData/CMakeLists.txt
@@ -141,6 +141,7 @@ include(AddProtoLibrary)
 add_proto_library(simDataProto STATIC
     simData
     NO_HEADER
+    APPEND_PATH
     PROTO_DIR GeneratedCode
     EXPORT_MACRO SIMDATAPROTO_EXPORT
 )
-- 
2.47.1.windows.2

